pwndbg/heap/ptmalloc.py

```diff
@@ -1741,31 +1781,17 @@ def thread_cache(self):
                     thread_cache_struct_addr = pwndbg.gdblib.memory.pvoid(
                         tls_base + self._thread_cache_offset
                     )
-                    if (
-                        pwndbg.gdblib.vmmap.find(thread_cache_struct_addr)
-                        and thread_cache_struct_addr in self.get_heap_boundaries()
-                    ):
+                    if pwndbg.gdblib.vmmap.find(thread_cache_struct_addr):
                         self._thread_cache = self.tcache_perthread_struct(thread_cache_struct_addr)
                         return self._thread_cache
 
             # If we still can't find the tcache, we guess tcache is in the first chunk of the heap
             # Note: The result might be wrong if the arena is being shared by multiple threads
             # And that's why we need to find the tcache address in TLS first
-            arena = self.get_arena()
-            heap_region = self.get_heap_boundaries()
+            arena = self.thread_arena or self.main_arena
```

Comment:
After the refactoring, we used many `allocator.thread_arena or allocator.main_arena` and `self.thread_arena or self.main_arena` in different places.

I think it's suitable if we have debug symbols, but if the target binary is single-threaded and doesn't have debug symbols, is it possible that we only use `main_arena` here?

Because the heuristic for `thread_arena` is more easily to fail than the heuristic of `main_arena`, and if it failed when resolving `thread_arena` here, we will get an`SymbolUnresolvableError` even if we are possible to get the correct result with `main_arena`.

So, how about we add a new property called `current_arena` for `DebugsymsHeap` and `HeuristicHeap`, and replace `allocator.thread_arena or allocator.main_arena` with `allocator.current_arena`?

Something like:

```python
class DebugSymsHeap(GlibcMemoryAllocator):
    @propery
    def current_arena(self):
        return self.thread_arena or self.main_arena
```

and

```python
class HeuristicHeap(GlibcMemoryAllocator):
    @propery
    def current_arena(self):
        if not self.multithreaded:
            # We return `main_arena` instead of `thread_arena` because `main_arena` is more reliable than `thread_arena` currently
            return self.main_arena
        return self.thread_arena
```

Reply:
Yes I agree that because `thread_arena` can fail, checking for a multi-threaded environment before attempting to resolve it is sensible. However, I removed the `get_arena()` and `get_heap_boundaries()` methods because they were ambiguous, and I think the idea of a "current arena" is still ambiguous...

What do you think to adding a check for `self.multithreaded` into `thread_arena`, then returning the main arena if the process is single-threaded? That way the name "thread_arena" is still accurate, since you'll still be getting thread 1's arena in a single-threaded environment.

Reply:
*Note to self
Address cases where malloc is initialized but the debugged thread is not attached to an arena.*

Reply:
> What do you think to adding a check for `self.multithreaded` into `thread_arena`, then returning the main arena if the process is single-threaded?

I think it's good, I agreed 👍

Btw, if we do this, we might need to modify the tests here: https://github.com/pwndbg/pwndbg/blob/5695658797d3284517ae4ca2ed3eeefafb9fbae1/tests/gdb-tests/tests/heap/test_heap.py#L366

to make sure the heuristic use `main_arena` correctly when single-threaded, and then assert `gdb.selected_thread().num == 2` and `*thread_arena != &main_arena` to check if the heuristic still works for multi-threaded.

(probably add a for-loop or something that we do the same tests but `continue` the process in the second iteration to enter the new thread)

Reply:
Good point.
My next PR focusses on writing more tests, I've noted this as one more to be added 👍